### PR TITLE
Fix negative hex test

### LINKED ISSUE #11968: Some riscv-compliance tests are not building at head with clang
`I-RF_x0-01` has trouble building with what looks like a sign extension issue:
`
ERROR: /usr/local/google/home/drewmacrae/opentitan/third_party/riscv-compliance/BUILD:9:21: Compiling riscv-test-suite/rv32i/src/I-RF_x0-01.S failed: (Exit 1): gcc failed:
 error executing command external/bazel_embedded/toolchains/lowrisc_toolchain_rv32imc/clang_wrappers/nix/gcc -MD -MF ... (remaining 82 argument(s) skipped)
                                                                                     
Use --sandbox_debug to see verbose messages from the sandbox                         
clang-11: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument]                                                                      
external/riscv-compliance/riscv-test-suite/rv32i/src/I-RF_x0-01.S:57:18: error: operand must be a symbol with %lo/%pcrel_lo/%tprel_lo modifier or an integer in the range [
-2048, 2047]                                                                         
    xori x0, x0, 0xFFFFF803               
                 ^        
`

this can be hit by checking out lowrisc/opentitan at 81513e and building
`bazel build //third_party/riscv-compliance:sim_verilator_I-RF_x0-01`

### PR DESCRIPTION
**Pull request by drewmacrae**
Patch, fix and re-enable I-RV_x0-01 test. Patch to record the resolution in our docs is also included, though also exists in a separate PR. 

Fixes #11968

### CODE REVIEW

**Comment by drewmacrae** [comment]
Fixes: #12170

File: doc/rm/asm_coding_style.md [hunk 1]
```diff
@@ -71,6 +71,15 @@ Example:
   or   a0, a0, 0x4
 ```
 
+To emit assembly that's compatible with [GNU "as", integer literals should be donated with a '-' prefix operator](https://sourceware.org/binutils/docs/as/Integers.html) like `-1`, or `-0x01`.
```
Comment (mcy) [inline doc/rm/asm_coding_style.md:74]: Is `donated` the word you meant here?

**Review by mcy — COMMENTED** [review]
Did we conclude this was _not_ an issue in the LLVM assembler? Seems like something we should ask upstream to fix, too, if so.

File: doc/rm/asm_coding_style.md (hunk 1 above)
Comment (drewmacrae) [reply doc/rm/asm_coding_style.md:74]: Nope!

**Comment by drewmacrae** [comment]
> Did we conclude this was not an issue in the LLVM assembler? Seems like something we should ask upstream to fix, too, if so.

TLDR, neither is wrong, but the documentation for the assembler is more clearly documented and correctly implemented. We should upstream the patch to the compliance tests.

There's a conflict between the assembler and the compliance test. The assembler's documented behavior is to demand negative numbers be expressed by a minus sign and then ignore two's complement representation of hex literals and throw a weird error message. The compliance test conforms to the RiscV assembly guide that states integers are two's complement and things that are enabled by that should be expected to work. It does so only in the case of some arithmetic instructions that don't require a unique signed implementation, and not regarding proper expression of literals. 

I was hoping to record the patch here and that it would get upstreamed at some point, (preferably without me learning about the process or timeline for doing so,) whether I can expect to get this patch approved in the meantime is a very real matter of preferences and best practices that I'll leave to the reviewers. I submitted it because I thought running a patched test and documenting the correct behavior was an improvement on skipping a test that's incompatible with our assembler.

**Comment by drewmacrae** [comment]
This was tracked and fixed upstream https://github.com/riscv-non-isa/riscv-arch-test/issues/139

**Comment by drewmacrae** [comment]
So I'll see if I can fix this by updating

File: sw/vendor/patches/riscv_compliance/0006-Fix-negative-immediate-in-assembly.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/riscv-test-suite/rv32i/src/I-RF_x0-01.S b/riscv-test-suite/rv32i/src/I-RF_x0-01.S
+index 8f282d502..c21d9d1a8 100644
+--- a/riscv-test-suite/rv32i/src/I-RF_x0-01.S
++++ b/riscv-test-suite/rv32i/src/I-RF_x0-01.S
+@@ -54,7 +54,7 @@ RV_COMPLIANCE_CODE_BEGIN
+     addi    x0, x0, 1
+     ori     x0, x0, 0x7F0
+     andi    x0, x0, 0x53F
+-    xori    x0, x0, 0xFFFFF803
++    xori    x0, x0, -0x7FD
```
Comment (luismarques) [inline sw/vendor/patches/riscv_compliance/0006-Fix-negative-immediate-in-assembly.patch:10]: This seems fine for now but if there's a fix upstream can't we just update the entire compliance suite?

File: doc/rm/asm_coding_style.md [hunk 3]
```diff
@@ -71,6 +71,15 @@ Example:
   or   a0, a0, 0x4
 ```
 
+To emit assembly that's compatible with [GNU "as": integer literals should be denoted with a '-' prefix operator](https://sourceware.org/binutils/docs/as/Integers.html) like `-1`, or `-0x01`.
+GNU "as" may flag two's complement representations of negative integers as out of range:
+```console
+error: operand must be a symbol with %lo/%pcrel_lo/%tprel_lo modifier or
+an integer in the range [-2048, 2047]
+    xori x0, x0, 0xFFFFF803
+                 ^
```
Comment (luismarques) [inline doc/rm/asm_coding_style.md:80]: Isn't this Clang's error message?

**Comment by luismarques** [comment]
> the RiscV assembly guide that states integers are two's complement and things that are enabled by that should be expected to work.

I'm not being able to find that. Could you please provide a link? @drewmacrae

File: sw/vendor/patches/riscv_compliance/0006-Fix-negative-immediate-in-assembly.patch (hunk 2 above)
Comment (drewmacrae) [reply sw/vendor/patches/riscv_compliance/0006-Fix-negative-immediate-in-assembly.patch:10]: I marked this as a draft in order to do that and I'll close this if I can get that to work.

File: doc/rm/asm_coding_style.md (hunk 3 above)
Comment (drewmacrae) [reply doc/rm/asm_coding_style.md:80]: Yeah, I can fix that.

**Comment by drewmacrae** [comment]
> the RiscV assembly guide that states integers are two's complement and things that are enabled by that should be expected to work.

> I'm not being able to find that. Could you please provide a link? @drewmacrae

Sorry, I recall it was WRT arithmetic not formatting of an immediate, and doesn't fully bear on the conversation, so I didn't cite and track the reference. I just mentioned it because it's the only thing I saw beyond the idea that the numbers are "hexadecimal" and I was trying to figure out whether to address efforts at fixing the toolchain or the tests and the documentation.

There's a discussion in the compliance test of why this occurs with 64 bit compilers that apparently interpret this as a positive value when a 32 bit compiler would read its intent correctly. Updating would keep our branch shorter and might bring in other fixes so that's my preference at the present.
